include/seastar/coroutine/exception.hh

```diff
@@ -68,7 +68,11 @@ namespace coroutine {
 /// ```
 struct exception {
     std::exception_ptr eptr;
-    explicit exception(std::exception_ptr eptr) : eptr(std::move(eptr)) {}
+
+    explicit exception(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
+
+    template <typename E>
+    explicit exception(E&& e) noexcept : eptr(std::make_exception_ptr(std::forward<E>(e))) {}
```

Comment:
It's a bad idea to allow construction from both exception_ptr and any type. A small mistake (like exception(&ep)) can complete change the behavior (seastar::make_exception_future has the same problem).

Let's use a different container for it.

Reply:
Or a helper function. In fact why not let the caller call make_exception_ptr?

Reply:
And looking at the patch changelog, perhaps delete make_exception signatures that take exception_ptr.

Reply:
One reason not to accept exception objects and not require the caller to call make_exception_ptr is to trace them,
as in https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/include/seastar/core/future.hh#L2087
Since they are never thrown (where they are traced by https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/src/core/exception_hacks.cc#L146).

We use a template instead of accept a reference to the std::exception in order to not lose the exception type when calling `std::make_exception_ptr`.  But I can add `requires std::derived_from<std::remove_reference_t<E>, std::exception>` to restrict the usage.

Reply:
@avikivity if we change the api to accept only exception_ptr or classes derived from std::exception, we'll need an api version change since this will break use cases like https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/tests/unit/smp_test.cc#L37-L59

Reply:
Keeping the restrictions only for coroutine::{return,make}_exception should be less of a problem since I image they have much fewer users.